File: server/audio_streamer.js
const fs = require('fs');
const path = require('path');

const CONTENT_TYPES = {
  ".mp3": "audio/mpeg",
  ".flac": "audio/flac",
  ".m4a": "audio/mp4",
  ".ogg": "audio/ogg",
  ".opus": "audio/opus",
  ".wav": "audio/wav",
};

function streamAudioFile(req, res, filePath) {
  if (!fs.existsSync(filePath)) {
    return res.status(404).json({ detail: "Audio file not found" });
  }

  const ext = path.extname(filePath).toLowerCase();
  const contentType = CONTENT_TYPES[ext] || "audio/mpeg";

  let stat;
  try {
    stat = fs.statSync(filePath);
  } catch (_) {
    return res.status(404).json({ detail: "Audio file cannot be accessed" });
  }

  const fileSize = stat.size;
  const range = req.headers.range;

  if (range) {
    let start = 0;
    let end = fileSize - 1;

    try {
      const parts = range.replace(/bytes=/, "").split("-");
      start = parseInt(parts[0], 10);
      if (parts[1]) {
        end = parseInt(parts[1], 10);
      }
      if (end >= fileSize) end = fileSize - 1;
    } catch (_) {
      start = 0;
      end = fileSize - 1;
    }

    if (start >= fileSize || start < 0) {
      res.writeHead(416, {
        "Content-Range": `bytes */${fileSize}`,
      });
      return res.end();
    }

    const chunkSize = (end - start) + 1;
    const fileStream = fs.createReadStream(filePath, { start, end });

    res.writeHead(206, {
      "Content-Range": `bytes ${start}-${end}/${fileSize}`,
      "Accept-Ranges": "bytes",
      "Content-Length": chunkSize,
      "Content-Type": contentType,
    });

    fileStream.pipe(res);
  } else {
    res.writeHead(200, {
      "Content-Length": fileSize,
      "Content-Type": contentType,
      "Accept-Ranges": "bytes",
    });

    fs.createReadStream(filePath).pipe(res);
  }
}

module.exports = {
  streamAudioFile
};

File: services/audio_streamer.py
import os
from pathlib import Path
from fastapi import HTTPException
from starlette.responses import StreamingResponse

def range_requests_response(request, file_path: str, content_type: str = "audio/mpeg"):
    path = Path(file_path)
    if not path.exists():
        raise HTTPException(status_code=404, detail="Audio file not found")
        
    file_size = path.stat().st_size
    range_header = request.headers.get("range")
    
    if range_header:
        # e.g., "bytes=0-1024"
        try:
            bytes_range = range_header.replace("bytes=", "").split("-")
            start = int(bytes_range[0])
            end = int(bytes_range[1]) if bytes_range[1] else file_size - 1
            if end >= file_size:
                end = file_size - 1
            chunk_size = (end - start) + 1
        except Exception:
            start = 0
            end = file_size - 1
            chunk_size = file_size
            
        def iterfile():
            with open(path, mode="rb") as f:
                f.seek(start)
                bytes_left = chunk_size
                while bytes_left > 0:
                    read_size = min(bytes_left, 64 * 1024)
                    data = f.read(read_size)
                    if not data:
                        break
                    bytes_left -= len(data)
                    yield data
                    
        headers = {
            "Content-Range": f"bytes {start}-{end}/{file_size}",
            "Accept-Ranges": "bytes",
            "Content-Length": str(chunk_size),
            "Content-Type": content_type,
        }
        return StreamingResponse(iterfile(), status_code=206, headers=headers)
        
    def iterfile_full():
        with open(path, mode="rb") as f:
            while chunk := f.read(64 * 1024):
                yield chunk
                
    headers = {
        "Accept-Ranges": "bytes",
        "Content-Length": str(file_size),
        "Content-Type": content_type,
    }
    return StreamingResponse(iterfile_full(), status_code=200, headers=headers)
